src/main/java/org/reactome/addlinks/db/ReferenceObjectCache.java

```diff
@@ -106,183 +106,217 @@ private static synchronized void buildReferenceCaches(String className, Map<Stri
 				{
-					logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.referenceDatabase, referenceObject);
-				}
-				catch (Exception e)
-				{
-					logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.identifier, referenceObject, e.getMessage());
+					logger.error("ArrayIndexOutOfBounds was caught! Identifier that triggered this was: {}; className: {}; objectCacheByIdentifier has {} items; objectCacheByIdentifier[identifier] has {} items", identifier, className, objectCacheByIdentifier.size(), objectCacheByIdentifier.containsKey(identifier) ? objectCacheByIdentifier.get(identifier).size() : " *THAT IDENTIFIER IS NOT IN THAT MAP!* ");
 					e.printStackTrace();
+					throw new Error(e);
 				}
+
+			}
+		}
+		catch (InvalidAttributeException e)
+		{
+			logger.error("Object {} does not have an identifier attribute, error: {}", referenceObject, e.getMessage());
+			e.printStackTrace();
+		}
+		catch (Exception e)
+		{
+			e.printStackTrace();
+			throw new Error(e);
+		}
+	}
+
+	/**
+	 * Adds an object to a cache that is indexed by ReferenceDatabase.
+	 * @param cacheByRefDB - the cache to add to.
+	 * @param referenceObject - the object to add.
+	 */
+	private static void addObjectToRefDBCache(Map<String, List<GKInstance>> cacheByRefDB, GKInstance referenceObject)
+	{
+		String refDBID;
+		try
+		{
+			refDBID = String.valueOf(((GKInstance) referenceObject.getAttributeValue(ReactomeJavaConstants.referenceDatabase)).getDBID());
+			if (! cacheByRefDB.containsKey(refDBID))
+			{
+				List<GKInstance> byRefDb = Collections.synchronizedList( new LinkedList<GKInstance>() );
+				byRefDb.add(referenceObject);
+				cacheByRefDB.put(refDBID, byRefDb);
+			}
+			else
+			{
+				cacheByRefDB.get(refDBID).add(referenceObject);
+			}
+		}
+		catch (InvalidAttributeException e)
+		{
+			logger.error("Could not get the \"{}\" attribute for {} because it not valid for this object.", ReactomeJavaConstants.referenceDatabase, referenceObject);
+		}
+		catch (Exception e)
+		{
+			logger.error("Could not get the \"{}\" attribute for {}. Reason: {}", ReactomeJavaConstants.identifier, referenceObject, e.getMessage());
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * Adds an object to a cache that is indexed by species.
+	 * @param className - the className of the object.
+	 * @param cacheBySpecies - the cache to add the object to.
+	 * @param referenceObject - the object to add.
+	 */
+	private static void addObjectToSpeciesCache(String className, Map<String, List<GKInstance>> cacheBySpecies, GKInstance referenceObject)
+	{
+		// Sets do not allow duplicates.
+		Set<String> allSpecies = null;
+		//ReferenceRNASequence objects don't always have Species info, for some reason, so we need to get that from the associated ReferenceGeneProduct
+		if (className.equals(ReactomeJavaConstants.ReferenceRNASequence))
+		{
+			Object species = null;
```

Comment:
Is there a reason this is an `Object` instead of a `GKInstance`?

Reply:
The code at this point doesn't use the _value_ of the `species` variable, it only is interested in whether or not a referenceObject _has_ a value for its `species` attribute. Since `getAttributeValue` returns `Object`, and since I don't actually _need_ to know anything more than is it null/non-null, I used `Object`. If the type of `species` is `GKInstance`, then I need to case the return of `getAttributeValue` to `GKInstance`, which seems silly since I'm not even going to use it as a `GKInstance`.

Reply:
True, but I think the cast is worth it because as a new reader of the code, I may have no idea what type species is if it is `Object` since that can hold any Java object!  For the sake of a cast (since getAttributeValue for species is guaranteed to return a GKInstance), I think it would help someone new to the code so they know its type.